src/abe/dippe.c

```diff
@@ -0,0 +1,613 @@
+    mpz_clear(dippe->p);
+    free(dippe->g1_A);
+    free(dippe->g1_UA);
+}
+
+/**
+ * Clears the public key and frees allocated memory
+ *
+ * @param pk A pointer to a cfe_dippe_pub_key struct
+ */
+void cfe_dippe_pub_key_free(cfe_dippe_pub_key *pk) {
+	free(pk->gt_alpha_A);
+	free(pk->g1_W_A);
+}
+
+/**
+ * Clears the secret key and frees allocated memory
+ *
+ * @param sk A pointer to a cfe_dippe_sec_key struct
+ */
+void cfe_dippe_sec_key_free(cfe_dippe_sec_key *sk) {
+	mpz_clear(sk->sigma);
+	cfe_vec_free(&(sk->alpha));
+	cfe_vec_free(&(sk->W));
+}
+
+/**
+ * Clears the ciphertext and frees allocated memory
+ *
+ * @param cipher A pointer to a cfe_dippe_cipher struct
+ */
+void cfe_dippe_cipher_free(cfe_dippe_cipher *cipher) {
+    free(cipher->C0);
+    free(cipher->Ci);
+}
+
+/**
+ * Clears the user secret key and frees allocated memory
+ *
+ * @param usk A pointer to a cfe_dippe_user_secret_key struct
+ */
+void cfe_dippe_user_secret_key_free(cfe_dippe_user_secret_key *usk) {
+    free(usk->Ki);
+}
+
+/**
+ * Initializes the DIPPE scheme and generates the global parameters
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param assump_size The size of the underlying assumption
+ */
+void cfe_dippe_setup(cfe_dippe *dippe, unsigned int assump_size) {
+
+    // temp vars
+    BIG_256_56 tmp_big;
+    mpz_t tmp_mpz;
+    mpz_init(tmp_mpz);
+
+    dippe->assump_size = assump_size;
+    mpz_init(dippe->p);
+    mpz_from_BIG_256_56(dippe->p, (int64_t*)CURVE_Order_BN254);
+
+    // A (k+1 x k)
+    cfe_vec A;
+    cfe_vec_init(&A, ((assump_size+1) * assump_size));
+    cfe_uniform_sample_vec(&A, dippe->p);
+
+    // g1^A (k+1 x k)
+    dippe->g1_A = (ECP_BN254*)cfe_malloc(sizeof(ECP_BN254) * (assump_size+1) * assump_size);
+    for (unsigned int i=0; i<(assump_size+1); i++) {
+        for (unsigned int j=0; j<assump_size; j++) {
+            ECP_BN254_generator(&(dippe->g1_A[(assump_size*i)+j]));
+            BIG_256_56_from_mpz(tmp_big, A.vec[(assump_size*i)+j]);
+            ECP_BN254_mul(&(dippe->g1_A[(assump_size*i)+j]), tmp_big);
+        }
+    }
+
+    // U (k+1 x k+1)
+    cfe_vec U;
+    cfe_vec_init(&U, ((assump_size+1) * (assump_size+1)));
+    cfe_uniform_sample_vec(&U, dippe->p);
+
+    // U(T)A  (k+1 x k)
+    cfe_vec UA;
+    cfe_vec_init(&UA, ((assump_size+1) * assump_size));
+    for (unsigned int i=0; i<(assump_size+1); i++) {
+        for (unsigned int j=0; j<assump_size; j++) {
+            mpz_set_ui(UA.vec[(assump_size*i)+j], 0);
+            for (unsigned int k=0; k<(assump_size+1); k++) {
+                mpz_mul(tmp_mpz, U.vec[i+(k*(assump_size+1))], A.vec[(assump_size*k)+j]);
+                mpz_add(UA.vec[(assump_size*i)+j], UA.vec[(assump_size*i)+j], tmp_mpz);
+                mpz_mod(UA.vec[(assump_size*i)+j], UA.vec[(assump_size*i)+j], dippe->p);
+            }
+        }
+    }
+
+    // g1^U(T)A (k+1 x k)
+    dippe->g1_UA = (ECP_BN254*)cfe_malloc(sizeof(ECP_BN254) * (assump_size+1) * assump_size);
```

Comment:
We have a struct representing a vector of `ECP_BN254` values called `cfe_vec_G1`. You could use this here. It would be even nicer if you defined a matrix of `ECP_BN254` values, say `cfe_mat_G1` and use it here instead.

Reply:
I replaced them and as you suggested, added additional structures for matrices of group elements.